sdk/python/kfp/dsl/for_loop.py

```diff
@@ -64,7 +73,7 @@ def _get_subvar_type(type_name: str) -> Optional[str]:
     return match['value_type'].lstrip().rstrip() if match else None
 
 
-class LoopArgument(pipeline_channel.PipelineParameterChannel):
+class LoopArgument(pipeline_channel.PipelineChannel):
```

Comment:
Would these implementations be simpler (even if longer) if we split `LoopArgument` into `class LoopArgumentParameter(pipeline_channel.PipelineParameterChannel)` and `class LoopArgumentArtifact(pipeline_channel.PipelineArtifactChannel)`?

I _think_ this would:
(a) Mirror the logical organization of `PipelineChannel`s
(b) Simplify the interfaces of `LoopArgumentParameter` and `LoopArgumentArtifact` compared to `LoopArgument`. No mixing of parameter and artifact concepts (as is the case with the `is_artifact_list` and `value` parameters).
(c) Eliminate changes needed in `pipeline_channel.py`
(d) Allow us to piggyback on more of the existing compilation logic in `pipeline_spec_builder.py`

I do agree with the intent of the current changes: a uniform mechanism for passing data, irrespective of data type, would be very nice to have. I think the complexity costs of this are somewhat high, however, since the rest of the KFP SDK internals don't follow this pattern beyond having a shared `PipelineChannel` ABC. As you've noticed, `PipelineChannel` and its subclasses don't have a great polymorphic relationship: `PipelineChannel` subclasses disregard the constructor of the `PipelineChannel` ABC, the Liskov Substitution Principle does not hold, etc. It's best to think of `PipelineChannel` as a form of "sentinel" parent class that merely informs code that an object that subclasses it is one of our several pipeline channel instances.

If we want to fix this pre-existing issue, I think it would make sense to address the inconsistency starting within `pipeline_channel.py`, though that's probably a bit ambitious for this one feature.

WDYT?

Reply:
I've made the changes in the other PR, yes I agree with (a) and (b) - the code is a little simpler by making the distinction between `LoopParameterArgument`and `LoopArtifactArgument`.

(c) and (d) are roughly the same